File: bench/raytrace/raytrace.go
// Raytrace head-to-head — Go reference port. Mirrors raytrace.ty operation-for-
// operation: struct-by-value Vec3, same shading math, same checksum fold.
// Go does not fuse float ops into FMA unless math.FMA is called explicitly, so
// the plain +/-/* here stay IEEE-strict like the C build's -ffp-contract=off.
package main

import (
	"fmt"
	"math"
)

type Vec3 struct{ x, y, z float64 }
type Sphere struct {
	center Vec3
	radius float64
	color  Vec3
}

func vadd(a, b Vec3) Vec3      { return Vec3{a.x + b.x, a.y + b.y, a.z + b.z} }
func vsub(a, b Vec3) Vec3      { return Vec3{a.x - b.x, a.y - b.y, a.z - b.z} }
func vscale(a Vec3, s float64) Vec3 { return Vec3{a.x * s, a.y * s, a.z * s} }
func vdot(a, b Vec3) float64   { return a.x*b.x + a.y*b.y + a.z*b.z }
func vlen(a Vec3) float64      { return math.Sqrt(vdot(a, a)) }
func vnorm(a Vec3) Vec3        { l := vlen(a); return vscale(a, 1.0/l) }

func hitSphere(o, d Vec3, s Sphere) float64 {
	oc := vsub(o, s.center)
	b := vdot(oc, d)
	c := vdot(oc, oc) - s.radius*s.radius
	disc := b*b - c
	if disc < 0.0 {
		return 0.0 - 1.0
	}
	t := 0.0 - b - math.Sqrt(disc)
	if t < 0.001 {
		return 0.0 - 1.0
	}
	return t
}

func clamp01(x float64) float64 {
	if x < 0.0 {
		return 0.0
	}
	if x > 1.0 {
		return 1.0
	}
	return x
}

func main() {
	nx, ny, ns := 1600, 1200, 40

	scene := make([]Sphere, 0, 64)
	for k := 0; k < ns; k++ {
		col := k % 5
		row := k / 5
		cx := (float64(col) - 2.0) * 0.9
		cy := 0.0 - 0.2
		cz := 0.0 - 2.5 - float64(row)*0.8
		rad := 0.35
		cr := float64((k*37)%100) / 100.0
		cg := float64((k*53)%100) / 100.0
		cb := float64((k*97)%100) / 100.0
		scene = append(scene, Sphere{Vec3{cx, cy, cz}, rad, Vec3{cr, cg, cb}})
	}
	scene = append(scene, Sphere{Vec3{0.0, 0.0 - 100.5, 0.0 - 3.0}, 100.0, Vec3{0.6, 0.6, 0.6}})

	light := vnorm(Vec3{0.0 - 1.0, 1.0, 0.5})
	origin := Vec3{0.0, 0.0, 0.0}

	var acc int64 = 0

	for j := 0; j < ny; j++ {
		for i := 0; i < nx; i++ {
			u := (float64(i)/float64(nx))*2.0 - 1.0
			v := 1.0 - (float64(j)/float64(ny))*2.0
			dir := vnorm(Vec3{u * 1.5, v, 0.0 - 1.0})

			best := 1000000.0
			hit := 0 - 1
			for m := 0; m < len(scene); m++ {
				t := hitSphere(origin, dir, scene[m])
				if t > 0.0 {
					if t < best {
						best = t
						hit = m
					}
				}
			}

			r, g, b := 0.0, 0.0, 0.0
			if hit < 0 {
				tt := 0.5 * (dir.y + 1.0)
				r = (1.0 - tt) + tt*0.5
				g = (1.0 - tt) + tt*0.7
				b = (1.0 - tt) + tt*1.0
			} else {
				s := scene[hit]
				p := vadd(origin, vscale(dir, best))
				n := vnorm(vsub(p, s.center))
				diff := vdot(n, light)
				if diff < 0.0 {
					diff = 0.0
				}
				sh := 0.15 + 0.85*diff
				r = s.color.x * sh
				g = s.color.y * sh
				b = s.color.z * sh
			}

			ir := int64(clamp01(r) * 255.0)
			ig := int64(clamp01(g) * 255.0)
			ib := int64(clamp01(b) * 255.0)
			acc = acc + ir + ig*7 + ib*13
		}
	}

	fmt.Printf("%d\n", acc)
}

File: bench/raytrace/raytrace.c
/* Raytrace head-to-head — C reference port. Mirrors raytrace.ty operation-for-
 * operation: same struct-by-value Vec3, same shading math, same checksum fold.
 * Build with -ffp-contract=off so no FMA fusion changes the bit pattern vs the
 * other ports (see run.sh). */
#include <stdio.h>
#include <math.h>

typedef struct { double x, y, z; } Vec3;
typedef struct { Vec3 center; double radius; Vec3 color; } Sphere;

static Vec3 vadd(Vec3 a, Vec3 b)   { return (Vec3){a.x + b.x, a.y + b.y, a.z + b.z}; }
static Vec3 vsub(Vec3 a, Vec3 b)   { return (Vec3){a.x - b.x, a.y - b.y, a.z - b.z}; }
static Vec3 vscale(Vec3 a, double s){ return (Vec3){a.x * s, a.y * s, a.z * s}; }
static double vdot(Vec3 a, Vec3 b) { return a.x * b.x + a.y * b.y + a.z * b.z; }
static double vlen(Vec3 a)         { return sqrt(vdot(a, a)); }
static Vec3 vnorm(Vec3 a)          { double l = vlen(a); return vscale(a, 1.0 / l); }

static double hit_sphere(Vec3 o, Vec3 d, Sphere s) {
    Vec3 oc = vsub(o, s.center);
    double b = vdot(oc, d);
    double c = vdot(oc, oc) - s.radius * s.radius;
    double disc = b * b - c;
    if (disc < 0.0) return 0.0 - 1.0;
    double t = 0.0 - b - sqrt(disc);
    if (t < 0.001) return 0.0 - 1.0;
    return t;
}

static double clamp01(double x) {
    if (x < 0.0) return 0.0;
    if (x > 1.0) return 1.0;
    return x;
}

int main(void) {
    int nx = 1600, ny = 1200, ns = 40;

    Sphere scene[64];
    int nsphere = 0;
    for (int k = 0; k < ns; k++) {
        int col = k % 5;
        int row = k / 5;
        double cx = ((double)col - 2.0) * 0.9;
        double cy = 0.0 - 0.2;
        double cz = 0.0 - 2.5 - (double)row * 0.8;
        double rad = 0.35;
        double cr = (double)((k * 37) % 100) / 100.0;
        double cg = (double)((k * 53) % 100) / 100.0;
        double cb = (double)((k * 97) % 100) / 100.0;
        scene[nsphere++] = (Sphere){{cx, cy, cz}, rad, {cr, cg, cb}};
    }
    scene[nsphere++] = (Sphere){{0.0, 0.0 - 100.5, 0.0 - 3.0}, 100.0, {0.6, 0.6, 0.6}};

    Vec3 light  = vnorm((Vec3){0.0 - 1.0, 1.0, 0.5});
    Vec3 origin = (Vec3){0.0, 0.0, 0.0};

    long long acc = 0;

    for (int j = 0; j < ny; j++) {
        for (int i = 0; i < nx; i++) {
            double u = ((double)i / (double)nx) * 2.0 - 1.0;
            double v = 1.0 - ((double)j / (double)ny) * 2.0;
            Vec3 dir = vnorm((Vec3){u * 1.5, v, 0.0 - 1.0});

            double best = 1000000.0;
            int hit = 0 - 1;
            for (int m = 0; m < nsphere; m++) {
                double t = hit_sphere(origin, dir, scene[m]);
                if (t > 0.0) {
                    if (t < best) { best = t; hit = m; }
                }
            }

            double r = 0.0, g = 0.0, b = 0.0;
            if (hit < 0) {
                double tt = 0.5 * (dir.y + 1.0);
                r = (1.0 - tt) + tt * 0.5;
                g = (1.0 - tt) + tt * 0.7;
                b = (1.0 - tt) + tt * 1.0;
            } else {
                Sphere s = scene[hit];
                Vec3 p = vadd(origin, vscale(dir, best));
                Vec3 n = vnorm(vsub(p, s.center));
                double diff = vdot(n, light);
                if (diff < 0.0) diff = 0.0;
                double sh = 0.15 + 0.85 * diff;
                r = s.color.x * sh;
                g = s.color.y * sh;
                b = s.color.z * sh;
            }

            long long ir = (long long)(clamp01(r) * 255.0);
            long long ig = (long long)(clamp01(g) * 255.0);
            long long ib = (long long)(clamp01(b) * 255.0);
            acc = acc + ir + ig * 7 + ib * 13;
        }
    }

    printf("%lld\n", acc);
    return 0;
}
